draft-ietf-masque-h3-datagram.md

```diff
@@ -117,26 +123,192 @@ following format (using the notation from the "Notational Conventions" section
+divisible by four.)
+
+Context ID:
+
+: A variable-length integer indicating the context ID of the datagram (see
+{{datagram-contexts}}).
+
+HTTP/3 Datagram Payload:
+
+: The payload of the datagram, whose semantics are defined by individual
+applications. Note that this field can be empty.
+
+Intermediaries parse the Quarter Stream ID field in order to associate the QUIC
+DATAGRAM frame with a stream. If an intermediary receives a QUIC DATAGRAM frame
+whose payload is too short to allow parsing the Quarter Stream ID field, the
+intermediary MUST treat it as an HTTP/3 connection error of type
+H3_GENERAL_PROTOCOL_ERROR. Intermediaries MUST ignore any HTTP/3 Datagram
+fields after the Quarter Stream ID.
+
+Endpoints parse both the Quarter Stream ID field and the Context ID field in
+order to associate the QUIC DATAGRAM frame with a stream and context within
+that stream. If an endpoint receives a QUIC DATAGRAM frame whose payload is too
+short to allow parsing the Quarter Stream ID field, the endpoint MUST treat it
+as an HTTP/3 connection error of type H3_GENERAL_PROTOCOL_ERROR. If an endpoint
+receives a QUIC DATAGRAM frame whose payload is long enough to allow parsing
+the Quarter Stream ID field but too short to allow parsing the Context ID
+field, the endpoint MUST abruptly terminate the corresponding stream with a
+stream error of type H3_GENERAL_PROTOCOL_ERROR.
+
+
+# CAPSULE HTTP/3 Frame Definition {#capsule-frame}
+
+CAPSULE allows reliably sending request-related information end-to-end, even in
+the presence of HTTP intermediaries.
+
+CAPSULE is an HTTP/3 Frame (as opposed to a QUIC frame) which SHALL only be sent
+in client-initiated bidirectional streams. Intermediaries MUST forward all
+received CAPSULE frame in their unmodified entirety on the same stream where it
+would forward DATA frames.
+
+This specification of CAPSULE currently uses HTTP/3 frame type 0xffcab5. If this
+document is approved, a lower number will be requested from IANA.
+
+~~~
+CAPSULE HTTP/3 Frame {
+  Type (i) = 0xffcab5,
+  Length (i),
+  Capsule Type (i),
+  Capsule Data (..),
+}
+~~~
+{: #capsule-frame-format title="CAPSULE HTTP/3 Frame Format"}
+
+The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
+The payload consists of:
+
+Capsule Type:
+
+: The type of this capsule.
+
+Capsule Data:
+
+: Data whose semantics depends on the Capsule Type.
+
+Endpoints which receive a Capsule with an unknown Capsule Type MUST silently
+drop that Capsule. Intermediaries MUST forward Capsules, even if they do not
+know the Capsule Type or cannot parse the Capsule Data.
+
+
+## The REGISTER_DATAGRAM_CONTEXT Capsule {#register-capsule}
+
+The REGISTER_DATAGRAM_CONTEXT capsule (type=0x00) allows an endpoint to inform
+its peer of the encoding and semantics of upcoming datagrams associated with
+a given context ID. Its Capsule Data field consists of:
+
+~~~
+REGISTER_DATAGRAM_CONTEXT Capsule {
```

Comment:
I consider that to be a feature rather than a bug. If we don't exercise this path, implementors will cut corners and we won't be able to deploy extensions down the road. Requiring the main path to use the extension mechanism ensures that the mechanism is exercised and functional.

Reply:
Most wg's approach grease by defining it and encouraging implementers to use it liberally enough to prevent this, but not require it.  HTTP/3 doesn't require a GREASE frame in order to function.

Reply:
This isn't about grease, it's about using your extension mechanism. HTTP/3 forces all HTTP implementations to correctly support HTTP headers, because otherwise they couldn't send or parse `:method`. That guarantees that everyone correctly handles headers. We could have said that if no HEADERS frame is sent then it means GET, but we decided against that.